pkg/secrets/passdriver/passdriver.go

```diff
@@ -0,0 +1,101 @@
+package passdriver
+
+import (
+	"bytes"
+	"context"
+	"io"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"sort"
+	"strings"
+
+	"github.com/pkg/errors"
+)
+
+// secretsPrefix is the key prefix of all secrets created through this driver
+var secretsPrefix = "container-secrets/"
+
+// errNoSecretData indicates that there is not data associated with an id
+var errNoSecretData = errors.New("no secret data with ID")
+
+// errNoSecretData indicates that there is secret data already associated with an id
+var errSecretIDExists = errors.New("secret data with ID already exists")
+
+// Driver is the passdriver object
+type Driver struct {
+	// passBinary is the binary to be used (either `pass` or `gopass`)
+	passBinary string
+}
+
+// NewDriver creates a new file driver.
+// rootPath is the directory where the secrets data file resides.
+func NewDriver(passBinary string) (*Driver, error) {
+	passdriver := new(Driver)
+	passdriver.passBinary = passBinary
+	return passdriver, nil
+}
+
+// List returns all secret IDs
+func (d *Driver) List() ([]string, error) {
+	out := &bytes.Buffer{}
+	if err := d.exec(context.TODO(), nil, out, "ls", secretsPrefix); err != nil {
+		return nil, err
+	}
+	var allID []string
+	for {
+		line, err := out.ReadString('\n')
+		if err != nil {
+			if err == io.EOF {
+				break
+			}
+			return nil, err
+		}
+		line = strings.Trim(line, "\n")
+		if len(line) > len(secretsPrefix) {
+			allID = append(allID, line[len(secretsPrefix):])
+		}
+	}
+	sort.Strings(allID)
+	return allID, nil
+}
+
+// Lookup returns the bytes associated with a secret ID
+func (d *Driver) Lookup(id string) ([]byte, error) {
+	out := &bytes.Buffer{}
+	key := filepath.Join(secretsPrefix, id)
+	if err := d.exec(context.TODO(), nil, out, key); err != nil {
+		return nil, err
+	}
+	if out.Len() == 0 {
+		return nil, errors.Wrapf(errNoSecretData, "%s", id)
+	}
+	return out.Bytes(), nil
+}
+
+// Store stores the bytes associated with an ID. An error is returned if the ID arleady exists
+func (d *Driver) Store(id string, data []byte) error {
+	in := bytes.NewReader(data)
+	key := filepath.Join(secretsPrefix, id)
+	if err := d.exec(context.TODO(), in, nil, "insert", key); err != nil {
+		return errors.Wrap(errSecretIDExists, id)
+	}
+	return nil
+}
+
+// Delete deletes the secret associated with the specified ID.  An error is returned if no matching secret is found.
+func (d *Driver) Delete(id string) error {
+	key := filepath.Join(secretsPrefix, id)
```

Comment:
We need to make sure the id passed in is just a valid id and not an escaped directory string.  i.e. something like "../../../etc" which would cause all of etc to be removed.

Reply:
this is now explicitly checked